SignalServiceKit/src/Storage/OWSPrimaryStorage.m

```diff
@@ -223,22 +180,48 @@ - (void)runSyncRegistrations
 - (void)runAsyncRegistrationsWithCompletion:(void (^_Nonnull)(void))completion
 {
     OWSAssert(completion);
+    OWSAssert(self.database);
 
     DDLogVerbose(@"%@ async registrations enqueuing.", self.logTag);
 
-    RunAsyncRegistrationsForStorage(self, ^{
-        OWSAssertIsOnMainThread();
-
-        OWSAssert(!self.areAsyncRegistrationsComplete);
-
-        DDLogVerbose(@"%@ async registrations complete.", self.logTag);
-
-        self.areAsyncRegistrationsComplete = YES;
-
-        completion();
-
-        [self verifyDatabaseViews];
-    });
+    // Asynchronously register other extensions.
+    //
+    // All sync registrations must be done before all async registrations,
+    // or the sync registrations will block on the async registrations.
+    [TSDatabaseView asyncRegisterThreadInteractionsDatabaseView:self];
+    [TSDatabaseView asyncRegisterThreadDatabaseView:self];
+    [TSDatabaseView asyncRegisterUnreadDatabaseView:self];
+    [self asyncRegisterExtension:[TSDatabaseSecondaryIndexes registerTimeStampIndex]
+                        withName:[TSDatabaseSecondaryIndexes registerTimeStampIndexExtensionName]];
+
+    [OWSMessageReceiver asyncRegisterDatabaseExtension:self];
+    [OWSBatchMessageProcessor asyncRegisterDatabaseExtension:self];
+
+    [TSDatabaseView asyncRegisterUnseenDatabaseView:self];
+    [TSDatabaseView asyncRegisterThreadOutgoingMessagesDatabaseView:self];
+    [TSDatabaseView asyncRegisterThreadSpecialMessagesDatabaseView:self];
+
+    [FullTextSearchFinder asyncRegisterDatabaseExtensionWithStorage:self];
+    [OWSIncomingMessageFinder asyncRegisterExtensionWithPrimaryStorage:self];
+    [TSDatabaseView asyncRegisterSecondaryDevicesDatabaseView:self];
+    [OWSDisappearingMessagesFinder asyncRegisterDatabaseExtensions:self];
+    [OWSFailedMessagesJob asyncRegisterDatabaseExtensionsWithPrimaryStorage:self];
+    [OWSIncompleteCallsJob asyncRegisterDatabaseExtensionsWithPrimaryStorage:self];
+    [OWSFailedAttachmentDownloadsJob asyncRegisterDatabaseExtensionsWithPrimaryStorage:self];
+    [OWSMediaGalleryFinder asyncRegisterDatabaseExtensionsWithPrimaryStorage:self];
+    [TSDatabaseView asyncRegisterLazyRestoreAttachmentsDatabaseView:self];
+
+    [self.database flushExtensionRequestsWithCompletionQueue:nil
```

Comment:
In this update YapDB introduced this method which we were approximating by having a completion handler on our last registration. 

I think this approach is a little more future proof.

Reply:
Great.  Since the stakes are high, I wonder how we can test this.  The cautious thing would be to add a `__block` flag that is set by the last view's completion block and is asserted by this methods' completion block, but that's probably a waste of time.